plugins/cuda/cuda_plugin.c

```diff
@@ -0,0 +1,406 @@
+		return false;
+	}
+
+	if (strstr(msg_buf, flag) == NULL) {
+		return false;
+	}
+
+	return true;
+}
+
+/* Retrieve the cuda restore thread TID from the root pid */
+static int get_cuda_restore_pid(int root_pid)
+{
+	char pid_buf[16];
+	char pid_out[CUDA_CKPT_BUF_SIZE];
+
+	snprintf(pid_buf, sizeof(pid_buf), "%d", root_pid);
+
+	const char *args[] = { CUDA_CHECKPOINT, "--get-restore-tid", "--pid", pid_buf, NULL };
+	int ret = launch_cuda_checkpoint(args, pid_out, sizeof(pid_out));
+	if (ret != 0) {
+		pr_err("Failed to launch cuda-checkpoint to retrieve restore tid: %s\n", pid_out);
+		return -1;
+	}
+
+	return atoi(pid_out);
+}
+
+static int cuda_process_checkpoint_action(int pid, const char *action, unsigned int timeout, char *msg_buf,
+					  int buf_size)
+{
+	char pid_buf[16];
+	char timeout_buf[16];
+
+	snprintf(pid_buf, sizeof(pid_buf), "%d", pid);
+
+	const char *args[] = { CUDA_CHECKPOINT, "--action", action, "--pid", pid_buf, NULL /* --timeout */,
+			       NULL /* timeout_val */, NULL };
+	if (timeout > 0) {
+		snprintf(timeout_buf, sizeof(timeout_buf), "%d", timeout);
+		args[5] = "--timeout";
+		args[6] = timeout_buf;
+	}
+
+	return launch_cuda_checkpoint(args, msg_buf, buf_size);
+}
+
+static int interrupt_restore_thread(int restore_pid)
+{
+	/* Since we resumed a thread that CRIU previously already froze we need to
+	 * INTERRUPT it once again, task was already SEIZE'd so we don't need to do
+	 * a compel_interrupt_task()
+	 */
+	if (ptrace(PTRACE_INTERRUPT, restore_pid, NULL, 0)) {
+		pr_err("Could not interrupt cuda restore tid %d after checkpoint, process may be in strange state\n",
+		       restore_pid);
+		return -1;
+	}
+
+	/* compel_wait_task() will fail here as expected but the work we need it to
+	 * do is actually done before the sigstop check that fails. Without this
+	 * compel_stop_on_syscall() will hang waiting for the task to hit a
+	 * sigreturn when cleaning up the parasite
+	 */
+	struct proc_status_creds creds;
+	if (compel_wait_task(restore_pid, -1, parse_pid_status, NULL, &creds.s, NULL)) {
+		pr_debug("wait_task failed as expected. ignoring\n");
+	}
+
+	return 0;
+}
+
+int cuda_plugin_checkpoint_devices(int pid)
+{
+	int restore_pid;
+	char msg_buf[CUDA_CKPT_BUF_SIZE];
+	int ret = 0;
+	int int_ret;
+
+	if (plugin_disabled) {
+		return 0;
+	}
+
+	restore_pid = get_cuda_restore_pid(pid);
+
+	/* We can possibly hit a race with cuInit() where we are past the point of
+	 * locking the process but at lock time cuInit() hadn't completed in which
+	 * case cuda-checkpoint will report that we're in an invalid state to
+	 * checkpoint
+	 */
+	if (restore_pid == -1) {
+		pr_info("No need to checkpoint devices on pid %d\n", pid);
+		return 0;
+	}
+
+	pr_info("Checkpointing CUDA devices on pid %d restore_pid %d\n", pid, restore_pid);
+	/* We need to resume the resume the checkpoint thread to prepare the
+	 * mappings for checkpointing
+	 */
+	if (ptrace(PTRACE_CONT, restore_pid, NULL, 0)) {
```

Comment:
we probably need to block signals before resuming a thread. You can look at parasite_run to figure out how it can be done.

Reply:
btw: we can consider adding compel calls to resume and suspend threads. It can be done in a separate pr.

Reply:
One more thing is here. If I am not mistaken, seccomp has been suspended to that moment (ptrace_suspend_seccomp). If it is right, we need to resume it before resuming the thread. Otherwise, we can have security concerns here.

Reply:
Fixed that up, let me know if it looks okay.